Add Migrate to Java 25 recipe

CodexRaunak opened the pull request:
Add a declarative recipe `MigrateToJava25` which has a recipe List of:
- `UpdateJenkinsfileForJava25` recipe - Update `Jenkinsfile` for Java 25 Testing  
- `SetupJenkinsfile` - make sure there is `Jenkinsfile `present as we need to upgrade it for Java 25
- Also we use open rewrite [UpgradeToJava25](https://docs.openrewrite.org/recipes/java/migrate/upgradetojava25)

### Testing done
`mvn clean install` 

### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [ ] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Ensure you have provided tests - that demonstrates feature works or fixes the issue

CodexRaunak commented:
TODO:
- Add pr title and body for this top level recipe
- Add testing unit + declarative 

I tried testing the `UpdateJenkinsfileForJava25` which made this [PR](https://github.com/Raunak80Madan/tuleap-oauth-plugin/pull/15/files)

jonesbusy commented:
Any reason to no use https://github.com/jenkins-infra/plugin-modernizer-tool/blob/main/plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/visitors/UpdateJenkinsFileVisitor.java?

There are already many use case tested: https://github.com/jenkins-infra/plugin-modernizer-tool/blob/d7cb7b18c0e0241087db6da7dd199023c1ba9686/plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/visitors/UpdateJenkinsFileVisitorTest.java#L23

jonesbusy reviewed (changes requested):
Check to use `UpdateJenkinsFileVisitor`

CodexRaunak commented:
> Any reason to no use https://github.com/jenkins-infra/plugin-modernizer-tool/blob/main/plugin-modernizer-core/src/main/java/io/jenkins/tools/pluginmodernizer/core/visitors/UpdateJenkinsFileVisitor.java?
> 
> 
> 
> There are already many use case tested: https://github.com/jenkins-infra/plugin-modernizer-tool/blob/d7cb7b18c0e0241087db6da7dd199023c1ba9686/plugin-modernizer-core/src/test/java/io/jenkins/tools/pluginmodernizer/core/visitors/UpdateJenkinsFileVisitorTest.java#L23
> 
> 
> 
> 

Oops, my bad, I will have a look.
Thanks

jonesbusy commented:
Yes that's good because we also ensure the Jenkinsfile follow the archetype as much as possible.

So if we can get current configuration (via metadata or scanning recipe) to ensure we append for now.

And when 17 will be gone from Jenkinsfile we can adapt the receipe to remove the 17 configuration

gounthar commented:
Sounds good to me.

CodexRaunak commented:
Rebased.
The following changes has been made:
- Update recipe `UpdateJenkinsfileForJava25` to collect the current configurations and then call the `UpdateJenkinsFileVisitor` with `java 25` added.
- Added unit tests for `UpdateJenkinsfileForJava25` and declarative tests for `MigrateToJava25`.
- Added `Java 25` as enum. 
- Added pr title and body for `MigrateToJava25`, let me know if there is any suggestions on improving description of the PR. Also made a test PR [here](https://github.com/Raunak80Madan/tuleap-oauth-plugin/pull/18)  

I noticed a thing, that the open rewrite [UpgradeToJava25](https://docs.openrewrite.org/recipes/java/migrate/upgradetojava25) recipe was giving validation error, recipe doesn't exist.
I also individually listed the recipes as [here](https://docs.openrewrite.org/recipes/java/migrate/upgradetojava25#recipe-source), but was giving the below error. Also all these recipes are declarative recipes defined in yaml files in open rewrite. Can we not use it?

```
[ERROR] Recipe validation error in org.openrewrite.java.migrate.UpgradeToJava25.recipeList[1] (in jar:file:///C:/Users/DELL%20PC/.m2/repository/org/openrewrite/recipe/rewrite-migrate-java/3.14.1/rewrite-migrate-java-3.14.1.jar!/META-INF/rewrite/java-version-25.yml): recipe 'org.openrewrite.java.migrate.RemoveSecurityAccessController' does not exist. 
2025-07-29T09:23:04.386Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] Recipe validation error in initialization: DeclarativeRecipe must not contain uninitialized recipes. Be sure to call .initialize() on DeclarativeRecipe. 
2025-07-29T09:23:04.386Z [INFO] [Thread=StreamPumper-systemOut] - i.j.t.p.core.impl.MavenInvoker # [ERROR] Recipe validation error in org.openrewrite.java.migrate.UpgradeToJava25.recipeList[1] (in jar:file:///C:/Users/DELL%20PC/.m2/repository/org/openrewrite/recipe/rewrite-migrate-java/3.14.1/rewrite-migrate-java-3.14.1.jar!/META-INF/rewrite/java-version-25.yml): recipe 'org.openrewrite.java.migrate.RemoveSecurityAccessController' does not exist. 
```

jonesbusy commented:
Thanks looks good.

I think missing recipe is only packaging/dependency issues

Possible exclusion issue: https://github.com/jenkins-infra/plugin-modernizer-tool/blob/main/plugin-modernizer-core/pom.xml#L147-L156
